torch_xla/experimental/xla_sharding.py

```diff
@@ -70,6 +71,134 @@ def get_logical_mesh(self):
     return self.device_ids.reshape(self.mesh_shape)
 
 
+class HybridMesh(Mesh):
+  ici_mesh_shape: Tuple[int, ...]
+  dcn_mesh_shape: Tuple[int, ...]
+
+  def __init__(self,
+               *,
+               ici_mesh_shape: Tuple[int, ...],
+               dcn_mesh_shape: Tuple[int, ...] = None,
+               axis_names: Tuple[str, ...] = None):
+    if dcn_mesh_shape == None:
+      dcn_mesh_shape = tuple([1] * len(ici_mesh_shape))
+    assert len(ici_mesh_shape) == len(dcn_mesh_shape)
+    mesh_shape = tuple([x * y for x, y in zip(ici_mesh_shape, dcn_mesh_shape)])
+    self.device_attributes = xr.global_device_attributes()
+    if 'slice_index' in self.device_attributes[0] and np.prod(
+        dcn_mesh_shape) == 1:
+      raise ValueError('Provide dcn_mesh_shape to create a mesh for multislice')
+    if 'slice_index' not in self.device_attributes[0] and np.prod(
+        dcn_mesh_shape) > 1:
+      raise ValueError('Invalid dcn_mesh_shape for single slice mesh')
+    self.ici_mesh_shape = ici_mesh_shape
+    self.dcn_mesh_shape = dcn_mesh_shape
+    if np.prod(dcn_mesh_shape) > 1 and 'slice_index' in self.device_attributes[
+        0]:  # multislice
+      mesh = self._create_hybrid_device_mesh(self.ici_mesh_shape,
+                                             self.dcn_mesh_shape)
+    else:
+      mesh = self._create_device_mesh(self.ici_mesh_shape)
+    device_ids = mesh.flatten()
+    super().__init__(device_ids, mesh_shape, axis_names)
+
+  def _get_physical_tpu_mesh(self, devices: Sequence[Any]) -> np.ndarray:
+    r"""Rearrange TPU devices in a slice into a physical mesh."""
+    assert xm.xla_device_hw(xm.xla_device()) == 'TPU'
+    device_coords = [self.device_attributes[d]['coords'] for d in devices]
```

Comment:
So, each coord is a list of the 3D coord of the device? Can you comment that in the code?